core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1278,12 +1322,19 @@ class KafkaController(val config: KafkaConfig,
       }
     }
 
-    info(s"Shutting down broker $id")
+    if (config.controlledShutdownSafetyCheckEnable && !safeToShutdown(id, brokerEpoch)) {
+      info(s"Controlled shutdown safety has prevented broker $id (broker epoch $brokerEpoch) from shutting down.")
+      throw new NotEnoughReplicasException(
+        s"Broker id $id cannot initiate shutdown without an impact on topic availability.")
+    }
 
     if (!controllerContext.liveOrShuttingDownBrokerIds.contains(id))
       throw new BrokerNotAvailableException(s"Broker id $id does not exist.")
 
-    controllerContext.shuttingDownBrokerIds.add(id)
+    info(s"Shutting down broker $id")
+
+    zkClient.recordBrokerShutdown(id, brokerEpoch, controllerContext.epochZkVersion)
```

Comment:
Is there a need to remove the /brokers/shutdown/<id> znode when the shutdown is complete?

Reply:
I have the same question. 
There are two cases: 
the shutting down broker comes back online. 
the shutting down broker is removed from the cluster. 

should the controller clean the znode in these two cases. 

when the controller detects a broker is offline, can the controller clear the shutting down broker node for it?
When a broker becomes the new controller, it can also clear shutting down broker znodes for nodes that were already offline?

Reply:
I don't think there is any need to remove these znodes, since they only serve as a marker that lets the controller always know the last epoch of a broker when it initiated shutdown. What is the concern with leaving these znodes as they are?

If you think it's important to remove them we can, we'd just have to evaluate when it is safe to do so. I don't think `/brokers/shutdown/<id>` when shutdown is complete is right, but I think you are right xiowu0 that removing when the controller detects that the broker is offline would work. It might be better to remove when the broker comes online with a larger epoch value and liveBrokerEpochs is updated.

Reply:
I think I liked just keeping a map of all shutdown epochs forever because it seemed simplest to get right, but if you feel strongly we should remove them we can figure that out.

Reply:
I pushed a change to remove these znodes in `kafka.controller.KafkaController#onBrokerStartup`